envoy/certificate_provider/certificate_provider.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/common/callback.h"
+#include "envoy/common/pure.h"
+#include "envoy/extensions/transport_sockets/tls/v3/cert.pb.h"
+#include "envoy/ssl/connection.h"
+
+#include "absl/strings/string_view.h"
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace CertificateProvider {
+
+class OnDemandUpdateMetadata {
```

Comment:
I'm not sure how to define this metadata and make it can be utilized by other cases easily.
Currently I only have [bumping](https://docs.google.com/document/d/1B6TaqkwjlUqIOYTLqcprwtsQsg8TXitLQQnvXg2WIbU/edit#) case, in which case I need to generate certs based on a real server cert, so I need to pass the information of real server cert to the cert provider instance. How can I make this interface generic?

Anybody has ideas?

Reply:
> @ggreenway @lizan @markdroth
> 
> Except the certificate provider instance management, I add the implementation for certificate provider plugin instance, including
> 
> ```proto
> message CertificateValidationContext {
>     ......
>     CertificateProviderPluginInstance ca_certificate_provider_instance
>     .....
> }
> ```
> 
> and
> 
> ```proto
> message CommonTlsContext {
>     ......
>     CertificateProviderPluginInstance tls_certificate_provider_instance
>     ......
> }
> ```
> 
> Besides, an interface for on-demand cert update request is added.
> 
> Please have a look.

@ggreenway @lizan @markdroth kindly ping. :)

Reply:
@lizan @markdroth Could you help review this?

Reply:
Can you elaborate how this callback will be used? It's not very clear from the design doc or this PR.

Reply:
`addOnDemandUpdateCallback` is used to add callback passed from worker thread.

Assuming we have a network filter to request cert upon data plane request, it needs to add a callback to cert provider and pass metadata to provider used for resource update,  cert provider will update the cert resources based on the metadata (on main thread since all config updates should be on main thread due to envoy design principle ). Once the update is done, the callback is invoked, it might contains some logic needs to be run on main thread and other logic wrapped in another nested callback needs to posted back to worker thread and resume the network filter chain.

The callback depends on feature. Above is just an example. But no matter what feature, we need pass metadata from data plane request to cert provider, this is why I was asking how to define this `OnDemandUpdateMetadata `

Reply:
The name is still very confusing :) I need a concrete example (or TLS bumping filter) to better understand this. But IIUC we need to think the threading model more carefully. It sounds similar to what [DNS cache](https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/dynamic_forward_proxy/dns_cache_impl.h) does in terms of threading model, so you might want to take a look on that.

Reply:
Which name are you confused with? Is it `OnDemandUpdateMetadata`?

Yes, it's similar to DNS cache, the same thing is we post the task to main thread to do DNS resolve or cert update, and callback to worker thread to continue the filter chain. For DNS, it passes the `host`  as a parameter to the function used for DNS resolve, here we need to pass `OnDemandUpdateMetadata`  as a parameter to the function used for cert generation. Since it's not just for bumping, we need it can cover all cases as much as possible. One alternative is listing all common config items as optional fields for cert generating in `OnDemandUpdateMetadata`, e.g. SANs, pkey type(RSA or ECDSA) etc. What do you think about this?

I think TLS bumping filter will not ready very soon, before it's ready, let's we think about if there is anything we can do for this PR?